File: Data Structures & Algorithms/binary-search/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l,r = 0, len(nums) - 1
        while l <= r:
            m_idx = l + (r - l) // 2
            m = nums[m_idx]
            if m == target: return m_idx
            if m < target:
                l = m_idx + 1
            else:
                r = m_idx - 1
        return -1

File: Data Structures & Algorithms/search-2d-matrix/submission-1.js
class Solution {
    /**
     * @param {number[][]} matrix
     * @param {number} target
     * @return {boolean}
     */
    searchMatrix(matrix, target) {
        if (!matrix) return false;

        let i = matrix[0].length -1 ;
        let j = 0;

        while (i >= 0 && j < matrix.length) {
            const smartPtr = matrix[j][i];
            if (target === smartPtr) return true;
            if (target < smartPtr) i--;
            else j++;
        }
        return false
    }
}
